# Noah Delano's Weekly Check-In: September 9th

### Problem I am proud with (maybe during office hours, we can discuss the proof further)

This week has been filled with a lot of confusion and re-learning as error boundaries is a topic that was always brushed over when learning about taylor polynomials, at least from what I remember, so it took a hot sec for me to fully comprehend what was going.

However, there was one problem I am especially proud of due to its applications to number theory, which was the stirling approximation: 

$$n! \approx n^{n}e^{-n}\sqrt{2\pi n}$$

I did some research on how the approx. was derived, but it is quite lengthy and there are many ways to derive, but to sum up really quickly, they basically used the equations:

$$ln(n!) = ln(n\cdot(n-1)\cdots2\cdot1)$$
$$ln(n!) = ln(n)+ln(n-1)+\cdots+ln(2)+ln(1)$$
$$ln(n!) = \sum\limits_{i=1}^n ln(n)$$
As well as Euler's Integeral:
$$n! = \int_{0}^\infty e^{-x}x^ndx$$

Using these equations, people are able to manipulate them to the stirling approx. which was really interesting. Here is a link to a good proof I found that I am too lazy to write down: https://csustan.csustan.edu/~tom/Lecture-Notes/Stirling/stirling-better.pdf

But besides the proof, I created a dataframe showing the many components associated with the approximation, including the absolute and relative error.

In [6]:
#Brings in my imports
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

In [15]:
factorial = 1 #Used as the falue that will be my factorial, beginning at 1!
df = pd.DataFrame() #Empty data frame

#empty lists for the various components I will collect
nlist = []
factorialList = []
st =[]
aError=[]
rError=[]

for n in range(1,21): # for n = 1,2,3,...,20
    
    #for every n-1, multiply by n
    factorial *= n
    
    #Striling approximation: n! = n^n e^−n √2πn
    stirling = np.sqrt(2.0*np.pi*n)*np.exp(-n)*(n**n)
    
    #Absolute difference between the real n! and the stirling approx. of n!
    absoluteError =  np.abs(factorial - stirling)
    
    #The absolute difference divided by the actual n!, or the relative error
    relativeError = absoluteError/ninitial
    
    nlist.append(n)
    st.append(stirling)
    factorialList.append(factorial)
    aError.append(absoluteError)
    rError.append(relativeError)

df['n value'] = nlist
df['n!'] = factorialList
df['Stirling Approx.'] = st
df['Absolute Error'] = aError
df['Relative Error'] = rError

df

n value                   n!  Stirling Approx.  Absolute Error  \
0         1                    1      9.221370e-01    7.786299e-02   
1         2                    2      1.919004e+00    8.099565e-02   
2         3                    6      5.836210e+00    1.637904e-01   
3         4                   24      2.350618e+01    4.938249e-01   
4         5                  120      1.180192e+02    1.980832e+00   
5         6                  720      7.100782e+02    9.921815e+00   
6         7                 5040      4.980396e+03    5.960417e+01   
7         8                40320      3.990240e+04    4.176045e+02   
8         9               362880      3.595369e+05    3.343127e+03   
9        10              3628800      3.598696e+06    3.010438e+04   
10       11             39916800      3.961563e+07    3.011749e+05   
11       12            479001600      4.756875e+08    3.314114e+06   
12       13           6227020800      6.187239e+09    3.978132e+07   
13       14          87178291200      8.666100e+10    5.172895e+08   
14       15        1307674368000      1.300431e+12    7.243646e+09   
15       16       20922789888000      2.081411e+13    1.086755e+11   
16       17      355687428096000      3.539483e+14    1.739099e+12   
17       18     6402373705728000      6.372805e+15    2.956908e+13   
18       19   121645100408832000      1.211128e+17    5.323138e+14   
19       20  2432902008176640000      2.422787e+18    1.011516e+16   

    Relative Error  
0     3.200416e-20  
1     3.329178e-20  
2     6.732306e-20  
3     2.029777e-19  
4     8.141849e-19  
5     4.078181e-18  
6     2.449921e-17  
7     1.716487e-16  
8     1.374131e-15  
9     1.237386e-14  
10    1.237925e-13  
11    1.362206e-12  
12    1.635139e-11  
13    2.126224e-10  
14    2.977368e-09  
15    4.466907e-08  
16    7.148251e-07  
17    1.215383e-05  
18    2.187979e-04  
19    4.157653e-03

Based on my dataframe, as n increases for n!, it appears the absolute error increases, which may seem bad at first, but you have to think that since the number is getting bigger, it doesn't mean the the percent difference is increasing as well. If you look at the relative error, it is decreasing. This means that as n increases, the percent decrease from n! to the approximate is getting smaller as the approx compared to the true value are pretty close to each other in terms of percentage. Therefore, the approximation is better as n increases.

<span style="color:red"> Some typos here and there, but certainly great content. I believe you went over the derivation in class? Great job Noah. </span>

### Problem I struggled with

From the homework, I struggled with #2 from section 1.2, where we had to find the 2nd degree approx. of $f(x) = e^xsin(x)$ about a=0, and bound the error in the approx. when $-\frac{\pi}{4} \leq x \leq \frac{\pi}{4}$.

Now finding the second degree approx. wasn't to bad as I first started off by finding the taylor polynomials of $e^x$ and $sin(x)$:

$e^x = 1 + x + \frac{x^2}{2!} + \cdots$

$sin(x) = x - \frac{x^3}{3!} + \cdots$

Therefore:

$e^xsin(x) = (1 + x + \frac{x^2}{2!})(x - \frac{x^3}{3!})$

$e^xsin(x) = x + x^2 + \cdots$

<span style="color:red"> Performing that termwise mutliplication actually wouldn't give you that polynomial.. you'd get $ x - \frac{x^4}{3!} + \cdots$, so I will take off some points for this. </span>

<span style="color:red"> -.5 pt </span>

Since we are only dealing with the second degree polynomial, we only need the 2nd power of the product of $e^xsin(x)$.

Thus:

$P_2(x) = x + x^2$

However, that is not what I struggled with. My trouble was in how to bound the error with $-\frac{\pi}{4} \leq x \leq \frac{\pi}{4}$.

My first thought process was trying to incorporate what was talked about in the video lectures. Since we are looking for the error, then we are looking at:

$f(x) - P_n(x) = R_n(x) = \frac{x^{n+1}}{(n+1)!}f^{(n)}(c_x)$  where  $0<c_x<x$

Since we are dealing with a second degree polynomial and $-\frac{\pi}{4} \leq x \leq \frac{\pi}{4}$, then:

$f(x) - P_2(x) = R_2(x) = \frac{x^3}{3!}f'''(c_x)$  where  $0<c_x<\frac{\pi}{4}$

However, the big problem with the work I have done is that the remainder is based of the remainder of only $e^x$ and not $e^xsin(x)$, which I didn't realize at first because I thought my issue was with the $c_x$, which did not look right to me.

So I kept going down this route as follows:

$f(x) = e^xsin(x)$

$f'(x) = e^x(sin(x)+cos(x))$

$f''(x) = 2e^xcos(x)$

$f'''(x) = 2e^x(cos(x)-sin(x))$

<span style="color:red"> This is really what you should have done for the first part. </span>

And inserted it into the remainder above:

$R_2(x) = \frac{x^3}{3!}2e^{c_x}(cos(c_x)-sin(c_x))$

$R_2(x) = \frac{x^3}{3}e^{c_x}(cos(c_x)-sin(c_x))$

But the $e^{c_x}$ just did not look right to me. So i talked to the professor, which helped immensly. Instead of looking at exactly what was done in the notes, I should look at $e^x$ and $sin(x)$ seperately using Theorem 1.2.1: Taylor's remainder:

$R_n(x) = \frac{(x-a)^{n+1}}{(n+1)!}f^{(n+1)}(c_x)$ where $\alpha\leq x\leq\beta$ with $c_x$ between $a$ and $x$.

Since we are dealing with second degree polynomials, then the taylor polynomial and the remainder would be:

$e^x \approx P_2(x) + R_2(x) = (1 + x + \frac{x^2}{2!}) + (\frac{x^3}{3!}e^{c_x})$

But since it is second degree, I would rather the remainder go down by one. So:

$e^x \approx (1 + x) + (\frac{x^2}{2!}e^{c_x})$

Using the same logic for the second degree polynomial and applying theorem 1.2.1 on $sin(x)$, then:

$sin(x) \approx (x) + ((-1)^3\frac{x^3}{3!}cos(c_x) = x - \frac{x^3}{3!}cos(c_x)$

<span style="color:red"> Hmmm, something about this step doesn't sit quite right with me. You're utizling the taylor about zero and yet using $cos(c_{x})$. I won't take off points for it but unless you show me more evidence I'm gonna advise you to go back to the drawing board on this one. </span>

Then similarly before we multiply them together:
    
$e^xsin(x) \approx (1 + x + \frac{x^2}{2!}e^{c_x})(x - \frac{x^3}{3!}cos(c_x))$

$e^xsin(x) \approx x + x^2 \frac{x^3}{6}e^{c_x} - \frac{x^3}{6}cos(c_x) - \frac{x^4}{6}cos(c_x) - \frac{x^5}{212}e^{c_x}cos(c_x)$

We want to know what the highest bound the error can be as it can be from 0 to some error value. So, now, we can isolate the remainder and try to find this higher bound:

$\frac{x^3}{6}e^{c_x} - \frac{x^3}{6}cos(c_x) - \frac{x^4}{6}cos(c_x) - \frac{x^5}{12}e^{c_x}cos(c_x)$

Since the boundary for $c_x$ is $0<c_x<1$, then the highest $cos(c_x)$ can be is $1$ if $c_x = 1$. So

$R_2(x) < \frac{x^3}{6}e^{c_x} - \frac{x^3}{6} - \frac{x^4}{6} - \frac{x^5}{12}e^{c_x}$

Then for $e^{c_x}$ to be an upper bound, then $c_x = 1$ as $e^0 < e^1$. So:

$R_2(x) < \frac{x^3}{6}e - \frac{x^3}{6} - \frac{x^4}{6} - \frac{x^5}{12}e$

Now that $c_x$ has been applied, lets look at the upper bounds for x, which has a boundary of $|x| \leq \frac{\pi}{4}$. Tp make this bound as high as possible, we want to get rid of the negatives as they will just make the bound lower, and therefore more inaccurate. So for odd functions, such as $x^3$, then we can say $x = \frac{\pi}{4}$, then:

$R_2(x) < \frac{x^3}{6}e - \frac{(-\frac{\pi}{4})^3}{6} - \frac{x^4}{6} - \frac{-(\frac{\pi}{4})^5}{12}e$

$R_2(x) < \frac{x^3}{6}e + \frac{(\frac{\pi}{4})^3}{6} - \frac{x^4}{6} + \frac{(\frac{\pi}{4})^5}{12}e$

But no matter if x is negative or positive, the even power will make the numerator positive, and therefore not allowing the negative sign to change. So the best we can do is let $x=0$ so that we won't subtract from our upper bound:

$R_2(x) < \frac{x^3}{6}e + \frac{(\frac{\pi}{4})^3}{6} - \frac{(0)^4}{6} + \frac{(\frac{\pi}{4})^5}{12}e$

$R_2(x) < \frac{x^3}{6}e + \frac{(\frac{\pi}{4})^3}{6} + \frac{(\frac{\pi}{4})^5}{12}e$

Then for the final $x^3$, since it is already positive, we can just put in the highest value of x, which is $\frac{\pi}{4}$. Therefore:

$R_2(x) < \frac{(\frac{\pi}{4})^3}{6}e + \frac{(\frac{\pi}{4})^3}{6} + \frac{(\frac{\pi}{4})^5}{12}e$

Now lets simplify:

$R_2(x) < (\frac{\pi}{4})^3(\frac{e}{6} + \frac{1}{6} + (\frac{\pi}{4})^2\frac{e}{12})$

$R_2(x) < (\frac{\pi}{4})^3(\frac{e}{6} + \frac{1}{6} + \frac{\pi^2e}{16\cdot12})$

$R_2(x) < (\frac{\pi}{4})^3[\frac{1}{6} + e(\frac{1}{6} + \frac{\pi^2}{192})]$

$R_2(x) < 0.3124643606$

Finally we have an upper bound that the 2nd degree polynomial error can possibly be. It was a lot of work and required some office hourse, bit finally I got an answer I am satisfied with.

<span style="color:red"> Points for justifying your weekly markdown problem. This is what I'd expect for something purely in latex. I'm not entirely sure about your logic in some of the bits but if you're able to show some more evidence to your point I'd say well done. </span>

Now when it came to group discussions, I definetly don't think I was as much of a contributor as I should have been. Since last Thursday, there has been a lot going on in my personal life, which prevented me from doing any of the assigned homework. With everything going on, I was very quiet Friday until I got a major panic attack, which lead me leaving class early. I was fine, but took the weekend to recooperate and began catching up on Monday where I tried answering and questions people had about bounds, but me knowledge was still limited as I was still catching up. But by wednesday, I was asking plenty of questions, especially on the terminology of the bounds, as well as how to isolate bounds on sertain inequalties of x, so I am quite proud of that. For the future, I just got to be able to get more time in studying the topics as I definetly still feel a bit behind everyone in this topic. I have a fundamental understanding of it, but getting a better understanding of error bounds is what I intend to improve on in the next few days. Hopefully, I will catch up soon.

<span style="color:red"> Life has a funny way of slapping us upside the head. Not to be ironic here but while this is certainly what we're looking for as far as self-reflection, I can't give you full points for 3 sections. </span>

<span style="color:red"> -1 pt </span>